Add DFTD3 and DFTD4 dispersion corrections

mtaillefumier opened the pull request:
The PR adds the London corrections implemented with the DFTD3 and DFTD4 frameworks using external libraries. Any comment improving the PR are welcome.

- [x] Custom D4 parameters
- [x] D3 correction
- [x] D3 custom parameters
- [x] D3 and D4 forces and stresses
- [x] D3 api fortran interface
- [x] D4 api fortran interface
- [x] vdw mpi

Commit:
Add DFTD3 dispersion correction

Commit:
Fix libvdwxc mpi support

toxa81 commented on cmake/modules/FindDFTD3.cmake:
```diff
@@ -0,0 +1,32 @@
+# find spglib
```
Not spglib, next line is also affected

toxa81 commented on cmake/modules/FindDFTD4.cmake:
```diff
@@ -0,0 +1,44 @@
+# find spglib
```
same here, not spglib

toxa81 commented on cmake/modules/FindDFTD4.cmake:
```diff
@@ -0,0 +1,44 @@
+# find spglib
+# if in non-standard location set environment variabled `SPG_DIR` to the root directory
```
and not SPG_DIR

mtaillefumier replied:
also fixed.

toxa81 commented on cmake/modules/FindDFTD4.cmake:
```diff
@@ -0,0 +1,44 @@
+endif()
+
+mark_as_advanced(
+  SIRIUS_DFTD3_FOUND
```
why DFTD3?

mtaillefumier replied:
typo. Fixed

toxa81 commented on src/api/sirius.f90:
```diff
@@ -7245,6 +7258,138 @@ subroutine sirius_set_atom_vector_field_aux(handler,ia,vector_field,error_code)&
+!> @brief Set the parameters controlling the dftd3 correction.
+!> @param [in] handler Simulation context handler.
+!> @param [in] method family of predefined parameters. Linked to the functional
+!> @param [in] damping__ damping correction, auto, manual.
```
There should be no tailing underscores

mtaillefumier replied:
fixed

toxa81 commented on src/api/sirius.f90:
```diff
@@ -7245,6 +7258,138 @@ subroutine sirius_set_atom_vector_field_aux(handler,ia,vector_field,error_code)&
+!> @param [in] alp alp parameter for dftd3 model.
+!> @param [in] beta beta parameter for dftd3 model.
+!> @param [out] error_code Error code.
+subroutine sirius_set_dftd3_correction(handler,method,damping__,atm,damping_term,&
```
no tailing underscores

toxa81 commented on src/context/input_schema.json:
```diff
@@ -907,6 +917,501 @@
+                        "hsesol",
+                        "lcwhpbe",
+                        "lcomegahpbe",
+                        "lcωhpbe",
```
UTF character. I don't know if this is a good idea

toxa81 commented on src/potential/dftd3_correction.cpp:
```diff
@@ -0,0 +1,207 @@
+#ifndef SIRIUS_USE_DFTD3
+    RTE_THROW("SIRIUS is compiled without dft-d3 support");
+#endif
+    if (!ctx_.cfg().parameters().dftd3_correction())
```
All `if` statements including single liners are still wrapped in braces {}

toxa81 commented on src/potential/dftd3_correction.cpp:
```diff
@@ -0,0 +1,207 @@
+                  "be set and match the XC functional. See the simple-dftd3 documentation for more details");
+    }
+
+    char* tmp = static_cast<char*>(std::malloc(sizeof(char) * 128));
```
no way we use malloc and memset in c++ code. Has to be changed to std::vector<char> and std::fill()

toxa81 commented on src/potential/dftd3_correction.cpp:
```diff
@@ -0,0 +1,207 @@
+        RTE_THROW(buffer_error);
+    }
+
+    std::free(tmp);
```
This to remove. Containers will free memory automatically.

toxa81 commented on src/potential/dftd3_correction.cpp:
```diff
@@ -0,0 +1,207 @@
+    if (!ctx_.cfg().parameters().dftd3_correction())
+        return;
+#ifdef SIRIUS_USE_DFTD3
+    auto lat_ = unit_cell_.lattice_vectors();
```
no tailing underscores in local variable names. Tailing underscores are only used for class members.

toxa81 commented on src/potential/dftd3_correction.hpp:
```diff
@@ -0,0 +1,111 @@
+    // energy
+    double energy_{0.0};
+
+    dftd3_error error_{nullptr};
```
Those seem to be pointers. Are they destroyed somewhere? Can they be wrapped into a smart pointer like std::unique_ptr? You can pass custom deallocator to a smart pointer.

mtaillefumier replied:
Yes. They are destroyed in dftd3.cpp. I moved the deleter to the header file. I was thinking about using `unique_ptr`  but the c api requires something like this `delete_error(&error_)` which will not work with `unique_ptr`. 

if we use unique_ptr here we have to do this for each of the pointer.

```
error_tmp = error_.release();
delete_error(&error_tmp);
```

I do not see any added advantage over a raw pointer as these pointers are only accessed by the class methods. The `libvdwxc` class does the same thing.

toxa81 commented on src/potential/dftd4_correction.cpp:
```diff
@@ -0,0 +1,155 @@
+    if (!ctx_.cfg().parameters().dftd4_correction())
+        return;
+#ifdef SIRIUS_USE_DFTD4
+    auto lat_ = unit_cell_.lattice_vectors();
```
No tailing underscores for all variables

toxa81 commented on src/potential/dftd4_correction.cpp:
```diff
@@ -0,0 +1,155 @@
+                  "be set and match the XC functional. See the simple-dftd4 documentation for more details");
+    }
+
+    char* tmp = static_cast<char*>(std::malloc(sizeof(char) * 128));
```
no C-style allocations, use containers instead

toxa81 commented on src/potential/potential.cpp:
```diff
@@ -148,6 +148,14 @@ Potential::Potential(Simulation_context& ctx__)
     }
 
+    if (ctx_.cfg().parameters().dftd3_correction()) {
+        dftd3_ = std::unique_ptr<dftd3>(new dftd3(ctx_, unit_cell_));
```
better std::make_unique<type>(constructor parameters...)

mtaillefumier replied:
fixed

toxa81 commented on src/potential/xc_functional.hpp:
```diff
@@ -37,6 +35,8 @@ class XC_functional : public XC_functional_base
 #if defined(SIRIUS_USE_VDWXC)
     vdwxc_data handler_vdw_{nullptr};
     bool vdw_functional_{false};
+    int func_      = -1;
```
no tailing underscores

mtaillefumier replied:
not sure to understand. it is a element of the class.

toxa81 replied:
agree, my bad

toxa81 reviewed (changes requested):
Two major complaints from my side
1) there should be no tailing underscores in local variables. This naming convention is reserved for class members
2) all if statements, including singe liners should still be wrapped in braces {}. I don't remember if our clang-formatter fixes this automatically

toxa81 commented:
+ gitlab ci/cd must pass

mtaillefumier committed:
Typos and stylistic changes

- Moved deleter to header files
- Cleaned up the cmake modules for dftd3 and dftd4
- Updated the ci/cd
- Fixed wrong DFLAGS
- Removed C like allocations -> replaced with std::string

Typos and stylistic changes

mtaillefumier commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
cscs-ci run default

simonpintarelli commented:
🤔 it fails on `sirius::multicharge` target not found. Curiously it doesn't fail at `find_package(dftd4 REQUIRED)`. The bits in the CI looks correct dftd3/4 are in the base uenv and +dftd3/4 variants are set in the correct places.

mtaillefumier commented:
> 🤔 it fails on `sirius::multicharge` target not found. Curiously it doesn't fail at `find_package(dftd4 REQUIRED)`. The bits in the CI looks correct dftd3/4 are in the base uenv and +dftd3/4 variants are set in the correct places.

That's strange. I will have a look thanks

Commit:
[spack] Add local version of dftd3 and dftd4 packages

mtaillefumier commented:
cscs-ci run default

Commit:
make pretty

mtaillefumier commented:
cscs-ci run default

Commit:
[spack] Add temporary spack recipe for spack 0.23.1

mtaillefumier commented:
cscs-ci run default

mtaillefumier commented:
cscs-ci run default

mtaillefumier committed:
Merge branch 'develop' into dft_d3

mtaillefumier commented:
cscs-ci run default

mtaillefumier commented:
cscs-ci run default

Commit:
make pretty

mtaillefumier commented:
cscs-ci run default

Commit:
Use the old package conventions

mtaillefumier commented:
cscs-ci run default

Commit:
variants wrongly placed

mtaillefumier commented:
cscs-ci run default

Commit:
Add missing dependency

mtaillefumier commented:
cscs-ci run default

Commit:
Fix initialization issue

mtaillefumier commented:
cscs-ci run default

toxa81 reviewed (commented):
LGTM
